go_covert_lib/controller/processor/symmetricEncryption/symmetricEncryption.go

```
@@ -0,0 +1,162 @@
+	case "Counter (CTR)":
+		cipherText = CTREncrypter(c.block, data, c.blockSize)
+	case "Output Feedback (OFB)":
+		cipherText = OFBEncrypter(c.block, data, c.blockSize)
+	default:
+		return nil, errors.New("Undefined mode selected")
+	}
+
+	return cipherText, nil
+}
+
+func CBCEncrypter(block cipher.Block, data []byte, blockSize int) []byte {
+	cipherText := make([]byte, blockSize+len(data))
+	iv := cipherText[:blockSize]
+
+	mode := cipher.NewCBCEncrypter(block, iv)
+	mode.CryptBlocks(cipherText[blockSize:], data)
+
+	return cipherText[:]
+}
+
+func CFBEncrypter(block cipher.Block, data []byte, blockSize int) []byte {
+	cipherText := make([]byte, blockSize+len(data))
+	iv := cipherText[:blockSize]
+
+	stream := cipher.NewCFBEncrypter(block, iv)
+	stream.XORKeyStream(cipherText[blockSize:], data)
+
+	return cipherText[:]
+}
+
+func CTREncrypter(block cipher.Block, data []byte, blockSize int) []byte {
+	cipherText := make([]byte, blockSize+len(data))
+	iv := cipherText[:blockSize]
+
+	stream := cipher.NewCTR(block, iv)
+	stream.XORKeyStream(cipherText[blockSize:], data)
+
+	return cipherText[:]
+}
+
+func OFBEncrypter(block cipher.Block, data []byte, blockSize int) []byte {
+	cipherText := make([]byte, blockSize+len(data))
+	iv := cipherText[:blockSize]
+
+	stream := cipher.NewOFB(block, iv)
+	stream.XORKeyStream(cipherText[blockSize:], data)
+
+	return cipherText[:]
+}
+
+func Pad(data []byte) []byte {
+	// padding is done by having 7 zero's at the start followed by the length 
+	// of the original message, the contents of the message, followed by a 
+	// series of zero's to fill the block size requirements of the symmetric  
+	// encryption algorithm 
+
+	b := make([]byte, 8)
+	b = append(b, data...)
+	binary.BigEndian.PutUint64(b[:8], uint64(len(data)))
+
+	for len(b)%aes.BlockSize != 0 {
+		b = append(b, 0)
+	}
+	return b
+}
+
+func UnPad(data []byte) []byte {
+	lenOfData := binary.BigEndian.Uint64(data[:8])
```

You cannot assume that the length you get here is correct. If the incoming data was not properly processed then this length could end up being larger than the data. You must check that the lenOfData is < (len(data) - 8)

---

As such, this function should return an error.